src/hotspot/os/linux/os_linux.cpp

```diff
@@ -2887,6 +2892,32 @@ void os::pd_free_memory(char *addr, size_t bytes, size_t alignment_hint) {
   }
 }
 
+static void warn_fail_pretouch_memory(void *start, void *end, size_t page_size,
+				      int err) {
+  warning("INFO: os::pretouch_memory(" PTR_FORMAT ", " PTR_FORMAT ", "
+          SIZE_FORMAT ") failed; error='%s' (errno=%d)",
+	  p2i(start), p2i(end), page_size,
+          os::strerror(err), err);
+}
+
+void os::pd_pretouch_memory(void *start, void *end, size_t page_size) {
+  char *cur = static_cast<char *>(start);
+  // Use madvise to pretouch on Linux first, and fallback to the generic method
+  // if unsupported. THP can form right after madvise rather than being
+  // assembled later.
+  if (::madvise(cur, static_cast<char *>(end) - cur, MADV_POPULATE_WRITE) == -1) {
+    int err = errno;
+    if (err == EINVAL) { // Not supported
+      // When using THP we need to always pre-touch using small pages as the OS
+      // will initially always use small pages.
+      page_size = UseTransparentHugePages ? (size_t)os::vm_page_size() : page_size;
```

Comment:
This assignment should be to a new variable named `pretouch_page_size` since it will only be used by `pretouch_memory_fallback`, and otherwise modifies the function parameter and also shadows the variable from `class os`.  Declaring it `const` would get you extra points from readers, but I accept that is not the style here.

Reply:
The code is moved from the Linux-specific code in pretouchTask.cpp. I think it would be fine here.

Reply:
The moved code was from a `#ifdef` block, which is why the overwriting of `page_size` was done. I prefer Peter's style suggestion of declaring a new `const` variable.